### PR TITLE
Carvelrepos

### PR DESCRIPTION
**Pull request by dlaloue-vmware**
implemented first version of the RepositoryService APIs.
this first version is enough to be able to support carvel package repositories such as the TCE package repository.

this provides partial implementation for issue #4681

the following features will come next:
 - support for Authentication
 - support for custom details for imgpkgbundle type
 - support for other repository types

### CODE REVIEW

**Comment by absoludity** [comment]
@dlaloue-vmware I [mentioned on Rafa's PR](https://github.com/vmware-tanzu/kubeapps/pull/4723#issuecomment-1136578096) that it'd be great, since you are both working on similar functionality in different plugins, to review each others' code here. I'm happy to provide input as well if required, but am keen for that to be a starting point. (I've been a bit of a bottle-neck for all changes lately, so am trying to reduce that).

**Comment by dlaloue-vmware** [comment]
@absoludity I agree and I will review his code, especially on the logic side.
but for both Rafa and I, this is really our first time writing go code, so i think it would be better to have someone with experience give a review as well.

**Comment by absoludity** [comment]
Yep, happy to do so :)

File: cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_data_adapters.go [hunk 1]
```diff
@@ -400,33 +411,207 @@ func (s *Server) buildPkgInstall(installedPackageName, targetCluster, targetName
...
+				Namespace: pr.Namespace,
+			},
+			Plugin:     GetPluginDetail(),
+			Identifier: pr.Name,
+		},
+		Name:            pr.Name,
+		NamespaceScoped: s.globalPackagingNamespace != pr.Namespace,
+	}
+
+	// handle fetch-specific configuration
```
Comment (gfichtenholt) [inline cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_data_adapters.go:432]: sorry this is confusing a bit. In create you have code to error if type != image bundle. 
here you have code to handle all kinds?

Comment (dlaloue-vmware) [reply cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_data_adapters.go:432]: I thought to make a difference between reading vs. writing.
if the repository exists in the cluster, there is no reason we cannot view it (at least part of it) and at least know of it existence.
but for creation/update, we cannot provide enough details, so creation/update is not supported.

Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_data_adapters.go:432]: maybe add a comment to that effect in the code?

**Review by gfichtenholt — APPROVED** [review]
I looked at all the files, see no glaring problems so far. There is a bit to do left, I understand, e.g. Auth. But so far, so good. 
My ask for the future is, please try and make the size of the PR smaller (e.g. don't do all operations+unit tests at once). Small, incremental, targeted check-ins are easier to review and find issues in, raher than several humongous files (I know the irony of me saying that, but @absoludity will back me on that)

File: cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_test.go [hunk 2]
```diff
@@ -6411,30 +6856,696 @@ func TestGetPackageRepositories(t *testing.T) {
...
+		requestCustomizer    func(request *corev1.GetPackageRepositoryDetailRequest) *corev1.GetPackageRepositoryDetailRequest
+		repositoryCustomizer func(repository *packagingv1alpha1.PackageRepository) *packagingv1alpha1.PackageRepository
+		responseCustomizer   func(response *corev1.GetPackageRepositoryDetailResponse) *corev1.GetPackageRepositoryDetailResponse
+		expectedStatusCode   codes.Code
+	}{
+		{
+			name: "not found",
+			repositoryCustomizer: func(repository *packagingv1alpha1.PackageRepository) *packagingv1alpha1.PackageRepository {
+				return repository
+			},
```
Comment (castelblanque) [inline cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_test.go:6927]: If there is a nil check for `repositoryCustomizer` on line 7022 there is no need to specify this?

**Comment by absoludity** [comment]
@dlaloue-vmware is a member of vmware-tanzu org. Removing cla-rejected label.

**Pull request merged**